### DISCUSSION TITLE
Libraries for zebra-scan RPC Interface

### DISCUSSION THREAD

**Discussion by mpguerra** [ZcashFoundation/zebra#8154] [Research]
Let's use this discussion to evaluate various different RPC libraries against our requirements.

---
### Security requirements

- Authenticates connections to make sure they are authorised to add or remove keys, or get results
- Encrypts connections to avoid leaking keys or results
- Requires proof of knowledge of keys before accessing results or deleting that key
- The client should authenticate the server before sending private viewing keys.

### Other requirements

- Can be used via Tor
---
Let's add each potential library as a new comment and add the evaluation against the criteria as comments under each library

**Comment by arya2** [comment]
I think [Tonic](https://docs.rs/tonic) is the best choice here:

> ### Security requirements
> - Authenticates connections to make sure they are authorised to add or remove keys, or get results

Auth can be handled with an [`Interceptor`](https://github.com/hyperium/tonic/blob/master/examples/src/interceptor/server.rs).

> - Encrypts connections to avoid leaking keys or results

It includes TLS via rustls.

> - Requires proof of knowledge of keys before accessing results or deleting that key

This could be checked in an interceptor for those routes.

> - The client should authenticate the server before sending private viewing keys.

The self-signed certificate needed for TLS should work here if I understand this correctly.

> ### Other requirements
> - Can be used via Tor

The provided transport can accept http1, and it should be possible to write our own server so it can be used via Tor.

https://docs.rs/tonic/latest/tonic/transport/server/struct.Server.html#method.accept_http1

#### Advantages

- Transport agnostic and could be extended to work with Tor
- Supports http1
- Encoding agnostic and could be extended to support JSON, rkyv, flatbuffers, or other encodings
- More popular and complete than `grpc` crate, good developer experience

#### Drawbacks

- Tonic only includes a `Codec` implementation for protobufs, supporting JSON may require us to implement a JSON codec.
- It seems it may need to be used with another version of [`Server`](https://docs.rs/tonic/latest/tonic/transport/struct.Server.html) so it can be used via Tor

**Reply by mpguerra** [reply]
Are there any other libraries that we've found that are not suitable?

**Reply by mpguerra** [reply]
I had a quick look and it does seem like `tonic` is the best bet.

The only thing that might not be straightforward is using it via Tor but we can investigate that later on.

**Reply by mpguerra** [reply]
@upbqdn @oxarbitrage Any other thoughts or shall we just go ahead and use `tonic`?

**Reply by oxarbitrage** [reply]
I think tonic is a good choice for our purposes. I think we can go ahead and use it.

**Reply by upbqdn** [reply]
I also think tonic is the only suitable option.

**Reply by arya2** [reply]
It seems [Solana uses tarpc](https://crates.io/crates/tarpc/reverse_dependencies), that could be another option. I'm not sure if it supports protobufs/prost, but it does seem to support bincode which looks comparable/slightly-better among compact formats [in these benchmarks](https://github.com/djkoloski/rust_serialization_benchmark).

**Comment by mpguerra** [comment]
@arya2 have you managed to do any additional research on whether we might want to use any other libraries for authentication?

**Reply by arya2** [reply]
I'm looking at [argon2](https://docs.rs/argon2/0.5.2/argon2/) for ~password~ key hashing ~and [jsonwebtoken](https://docs.rs/jsonwebtoken/latest/jsonwebtoken/) for signing cookies/tokens~.

**Reply by upbqdn** [reply]
We could use one of these https://github.com/RustCrypto/password-hashes for passwords, argon2 is among them.

**Comment by mpguerra** [comment]
We're not going to do any further work on the scanner
